src/libostree/ostree-repo.c

```diff
@@ -156,6 +156,513 @@ G_DEFINE_TYPE (OstreeRepo, ostree_repo, G_TYPE_OBJECT)
 
 #define SYSCONF_REMOTES SHORTENED_SYSCONFDIR "/ostree/remotes.d"
 
+/* Repository locking
+ *
+ * To guard against objects being deleted (e.g., prune) while they're in
+ * use by another operation is accessing them (e.g., commit), the
+ * repository must be locked by concurrent writers.
+ *
+ * The locking is implemented by maintaining a thread local table of
+ * lock stacks per repository. This allows thread safe locking since
+ * each thread maintains its own lock stack. See the OstreeRepoLock type
+ * below.
+ *
+ * The actual locking is done using either open file descriptor locks or
+ * flock locks. This allows the locking to work with concurrent
+ * processes. The lock file is held on the ".lock" file within the
+ * repository.
+ *
+ * The intended usage is to take a shared lock when writing objects or
+ * reading objects in critical sections. Exclusive locks are taken when
```

Comment:
Yeah, it's something I punted on because the semantics are a little weird. Internally, ostree reads it into memory once and then it's accessed via `ostree_repo_get_config` or `ostree_repo_copy_config`. I don't think locking there would be helpful.

What I think you could do is take a shared lock during `ostree_repo_reload_config` (when the files are actually read) and take an exclusive lock during `ostree_repo_write_config` and `ostree_repo_remote_{add,change,delete}`. But then I think it can't really be managed all the way internally because the typical pattern is `ostree_repo_copy_config`, update keyfile, `ostree_repo_write_config`. You'd really want to take an exclusive lock around the whole set so someone else doesn't change it while you're making keyfile modifications. That's certainly something flatpak does. But I didn't think all the way through it so punted for now.

Reply:
Oh, and `ostree_repo_reload_config` is called during `ostree_repo_open`, so I'd be worried about locking before you could even open the repo. But I guess you could probably skip the locking there if `self->inited` is `FALSE`.

Reply:
Seems reasonable to me to punt on this for now, but please file a follow-up issue about it.

Reply:
Will address after #1343.